docker/lib/dependabot/docker/file_updater.rb

```diff
@@ -47,33 +52,40 @@ def updated_dependency_files
 
       private
 
+      sig { returns T.nilable(Dependabot::Dependency) }
       def dependency
         # Dockerfiles will only ever be updating a single dependency
         dependencies.first
       end
 
+      sig { override.void }
       def check_required_files
         # Just check if there are any files at all.
         return if dependency_files.any?
 
         raise "No Dockerfile!"
       end
 
+      sig { params(file: Dependabot::DependencyFile).returns(String) }
       def updated_dockerfile_content(file)
         old_sources = previous_sources(file)
         new_sources = sources(file)
 
-        updated_content = file.content
+        updated_content = T.let(file.content, T.untyped)
 
         old_sources.zip(new_sources).each do |old_source, new_source|
-          updated_content = update_digest_and_tag(updated_content, old_source, new_source)
+          updated_content = update_digest_and_tag(updated_content, old_source, T.must(new_source))
         end
 
         raise "Expected content to change!" if updated_content == file.content
 
         updated_content
       end
 
+      sig do
+        params(previous_content: String, old_source: T::Hash[Symbol, T.nilable(String)],
+               new_source: T::Hash[Symbol, T.nilable(String)]).returns(String)
```

Comment:
Can we have a type alias for this? `T::Hash[Symbol, T.nilable(String)]`

Tagging @JamieMagee in case this already exists

Reply:
We don't currently have an alias for it.

Sometimes we use `T::Hash[Symbol, T.nilable(String)]`  and sometimes we use `T::Hash[Symbol, T.untyped]` depending on the exact key we extract from the `Hash`. If we absolutely know it's going to be a String, we can use the slightly stricter version, but if we're extracting multiple keys, or nested keys in the same method, `T.untyped` is the best approach right now.

In the long term, we should have a class that defines all the possible properties this can have. Similar to what we did for `Credential` in #8967.

Reply:
Thank you so much for clarifying.